### PR TITLE
BERT pretraining

### PR DESCRIPTION
**Pull request by ZiyueHuang**
## Description ##
@sxjscience 

### Essentials ###
- [ ] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

cc @dmlc/gluon-nlp-team

### CODE REVIEW

File: scripts/pretraining/bert/create_pretraining_data.py [hunk 1]
```diff
@@ -0,0 +1,532 @@
...
+    """A single training instance (sentence pair)."""
+
+    def __init__(self, tokens, segment_ids, masked_lm_positions,
+                 masked_lm_labels, is_random_next, vocab):
+        self.tokens = tokens
+        self.segment_ids = segment_ids
+        self.is_random_next = is_random_next
+        self.masked_lm_positions = masked_lm_positions
+        self.masked_lm_labels = masked_lm_labels
+        self.vocab = vocab
```
Comment (sxjscience) [inline scripts/pretraining/bert/create_pretraining_data.py:41]: May consider to remove the vocab from the TrainingInstance. You may also try to make it a dataclass (See SQuAD example: https://github.com/dmlc/gluon-nlp/blob/6c1ea59390772c0d9feac7d8fb03381d5db6d91c/scripts/question_answering/squad_utils.py#L59-L93)

File: scripts/pretraining/bert/create_pretraining_data.py [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+    (2) Blank lines between documents. Document boundaries are needed so
+    that the "next sentence prediction" task doesn't span between documents.
+
+    The function expect arguments packed in a tuple as described below.
+
+    Parameters
+    ----------
+    input_files : list of str
+        List of paths to input text files.
+    tokenizer : BERTTokenizer
```
Comment (sxjscience) [inline scripts/pretraining/bert/create_pretraining_data.py:204]: Should be a BaseTokenizer object.

File: scripts/pretraining/bert/create_pretraining_data.py [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+    assert len(masked_lms) <= num_to_predict
+    masked_lms = sorted(masked_lms, key=lambda x: x.index)
+
+    masked_lm_positions = []
+    masked_lm_labels = []
+    for p in masked_lms:
+        masked_lm_positions.append(p.index)
+        masked_lm_labels.append(p.label)
+
+    return (output_tokens, masked_lm_positions, masked_lm_labels)
```
Comment (sxjscience) [inline scripts/pretraining/bert/create_pretraining_data.py:513]: No need to wrap them inside a tuple.

File: scripts/pretraining/bert/pretraining_utils.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+        super(MaskedAccuracy, self).__init__(name, axis=axis,
+            output_names=output_names, label_names=label_names)
+        self.axis = axis
+
+    def update(self, labels, preds, masks=None):
+        masks = [None] * len(labels) if masks is None else masks
+        for label, pred_label, mask in zip(labels, preds, masks):
+            if pred_label.shape != label.shape:
+                #pred_label = pred_label.argmax(axis=self.axis)
+                pred_label = mx.npx.topk(pred_label.astype('float32', copy=False),
```
Comment (sxjscience) [inline scripts/pretraining/bert/pretraining_utils.py:260]: Have you met some bugs of argmax?

File: scripts/pretraining/bert/run_pretraining.py [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+                                 warmup_begin_lr=0,
+                                 pwr=1,
+                                 final_lr=0,
+                                 warmup_steps=warmup_steps,
+                                 warmup_mode='linear')
+    optimizer_params = {'learning_rate': args.lr,
+                        'wd': args.wd,
+                        'lr_scheduler': lr_scheduler,
+                        }
+    if args.optimizer == 'adamw':
```
Comment (sxjscience) [inline scripts/pretraining/bert/run_pretraining.py:235]: We can choose to use lamb by default: https://mxnet.apache.org/versions/1.6/api/python/docs/api/optimizer/index.html#mxnet.optimizer.LAMB

File: scripts/pretraining/bert/create_pretraining_data.py [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+                    a_end = random.randint(1, len(current_chunk) - 1)
+
+                tokens_a = []
+                for j in range(a_end):
+                    tokens_a.extend(current_chunk[j])
+
+                tokens_b = []
+                # Random next
+                is_random_next = False
+                if len(current_chunk) == 1 or random.random() < 0.5:
```
Comment (sxjscience) [inline scripts/pretraining/bert/create_pretraining_data.py:371]: You may try to implement the sentence order prediction objective in ALBERT. See https://github.com/google-research/albert/blob/cbce824721148a8eb2621f1b1c1af252ac5121cc/create_pretraining_data.py#L338-L345

File: scripts/pretraining/bert/pretraining_utils.py [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+        lengths = dataset.get_field('valid_lengths')
+    else:
+        lengths = dataset.transform(lambda input_ids, segment_ids, masked_lm_positions, \
+                                           masked_lm_ids, masked_lm_weights, \
+                                           next_sentence_labels, valid_lengths: \
+                                        valid_lengths, lazy=False)
+    sampler = FixedBucketSampler(lengths,
+                                 batch_size=batch_size,
+                                 num_buckets=num_buckets,
+                                 ratio=0,
```
Comment (sxjscience) [inline scripts/pretraining/bert/pretraining_utils.py:76]: If it's 0, is it equivalent to BatchSampler?

File: scripts/pretraining/bert/pretraining_utils.py [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+
+    Parameters
+    ----------
+    batch_size : int
+        The batch size per GPU.
+    shuffle : bool
+        Whether to shuffle the data.
+    num_buckets : int
+        The number of buckets for the FixedBucketSampler for training.
+    vocab : BERTVocab
```
Comment (sxjscience) [inline scripts/pretraining/bert/pretraining_utils.py:98]: Change to Vocab.

File: scripts/pretraining/bert/create_pretraining_data.py (hunk 1 above)
Comment (ZiyueHuang) [reply scripts/pretraining/bert/create_pretraining_data.py:41]: This is used for debugging (transforming the token id to the actual words in its string representation).

File: scripts/pretraining/bert/create_pretraining_data.py (hunk 2 above)
Comment (ZiyueHuang) [reply scripts/pretraining/bert/create_pretraining_data.py:204]: done

File: scripts/pretraining/bert/create_pretraining_data.py (hunk 3 above)
Comment (ZiyueHuang) [reply scripts/pretraining/bert/create_pretraining_data.py:519]: This is needed for multiprocessing (worker_pool.apply).

File: scripts/pretraining/bert/pretraining_utils.py (hunk 4 above)
Comment (ZiyueHuang) [reply scripts/pretraining/bert/pretraining_utils.py:260]: `mx.npx.topk` is faster than `mx.np.argmax` about 50 times. The issue (https://github.com/apache/incubator-mxnet/issues/11061) still exists in MXNet 2.0.

```
import time
import mxnet as mx
import numpy as np
import os
tmp = mx.np.random.normal(-1, 1, (64,300000), ctx=mx.gpu())
tic = time.time()
for i in range(20):
   if (i == 5):
        begin = time.time();
   elif (i == 15):
        end = time.time();
   tic = time.time()
   #out = mx.np.argmax(tmp, axis=1)
   out = mx.npx.topk(tmp, k=1, ret_typ='indices', axis=1, dtype=np.int32)
   out.wait_to_read()
   toc = time.time() - tic
   print ("used time %f:"%toc)
avg = (end - begin) / 10
print ("avg time %f"%avg)
```

File: scripts/pretraining/bert/run_pretraining.py (hunk 5 above)
Comment (ZiyueHuang) [reply scripts/pretraining/bert/run_pretraining.py:241]: lamb suites for large batch size.

File: scripts/pretraining/bert/create_pretraining_data.py (hunk 6 above)
Comment (ZiyueHuang) [reply scripts/pretraining/bert/create_pretraining_data.py]: done

File: scripts/pretraining/bert/pretraining_utils.py (hunk 7 above)
Comment (ZiyueHuang) [reply scripts/pretraining/bert/pretraining_utils.py:76]: Not sure. I just follow the electra and the original bert.

File: scripts/pretraining/bert/pretraining_utils.py (hunk 8 above)
Comment (ZiyueHuang) [reply scripts/pretraining/bert/pretraining_utils.py]: done

File: scripts/pretraining/bert/pretraining_utils.py (hunk 4 above)
Comment (sxjscience) [reply scripts/pretraining/bert/pretraining_utils.py:260]: Thanks Ziyue, I can reproduce.... 

@szha Our argmax is too slow...

Comment (sxjscience) [reply scripts/pretraining/bert/pretraining_utils.py:260]: I'm actually not aware of this. Argmax is quite easy to accelerate and we can even call packages like cub.

**Review by sxjscience — APPROVED** [review]

**Comment by sxjscience** [comment]
LGTM overall. But we may later need to consider to merge the Electra + Bert (nsp + mlm) + Albert (sop + mlm) implementations.

**Comment by sxjscience** [comment]
Would you try to merge the upstream/master? It seems that there's something wrong in the GPU test.

**Comment by ZiyueHuang** [comment]
> Would you try to merge the upstream/master? It seems that there's something wrong in the GPU test.

done

**Pull request merged**